rpcserver.go

```diff
@@ -3237,20 +3246,41 @@ func (r *rpcServer) sendPayment(stream *paymentStream) error {
 		}
 	}()
 
+sendLoop:
 	for {
 		select {
+
+		// If we encounter and error either during sending or
+		// receiving, we return directly, clsoing the stream.
 		case err := <-errChan:
 			return err
 
-		case payIntent := <-payChan:
+		case <-r.quit:
+			return errors.New("rpc server shutting down")
+
+		case payIntent, ok := <-payChan:
+			// If the receive loop is done, we break the send loop
+			// and wait for the ongoing payments to finish before
+			// exiting.
+			if !ok {
+				break sendLoop
+			}
+
 			// We launch a new goroutine to execute the current
 			// payment so we can continue to serve requests while
 			// this payment is being dispatched.
+			wg.Add(1)
 			go func() {
+				defer wg.Done()
+
 				// Attempt to grab a free semaphore slot, using
 				// a defer to eventually release the slot
 				// regardless of payment success.
-				<-htlcSema
+				select {
+				case <-htlcSema:
+				case <-reqQuit:
```

Comment:
doesn't seem like this case can ever be executed, since `close(reqQuit)` is deferred but now we wait for the goroutines to exit. i think this is fixed by deferring `wg.Wait()` before deferring the close

Reply:
If I understand this correctly, there's a return on the main execution line outside of the goroutines (line 3256 in the current PR state). That return indicates an error and signals all gorotutines to quit as soon as possible.

Reply:
Yeah, the waitgroup is only being waited on in the non-error case, where there are no payments left to send. This handle the error cases where we return early.

Reply:
got it thanks for the explanation, makes sense